Social Icons: Add icon & background color options

aaronrobertshaw opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/21605

## Description

This is an alternate approach to https://github.com/WordPress/gutenberg/pull/25372 using CSS variables on the primary Social Links block that are then used by styles on the individual Social Link blocks. 

Compared to the original approach this PR:
- Avoids having to maintain and pass color hex values through context to the individual blocks
- Does not need to manually adjust server side rendering of social link blocks
- Social Links block gets icon and background colors set as CSS variables which are used by CSS rules targeting the individual Social Link blocks.

## How has this been tested?

Manual Testing.

1. Create new post and add social links block with a few links added to it.
2. Save post
3. Checkout this PR
4. Reload the editor and ensure no block validation errors
5. Select the social links block and toggle icon or background color options
6. Try switching block styles
    * "Logos only" style will clear any previous background color selection
    * Having only an icon color selected will result in no icon background color
7. Save the post and confirm colors on the frontend match the editor.

## Screenshots
| TwentyTwentyOne | TwentyTwenty |
|---|---|
| ![SocialLinksCSSVariables](https://user-images.githubusercontent.com/60436221/104686498-0e5ae880-5749-11eb-9a62-c3426cc1600d.gif) | ![SocialLinksCSSVariables2](https://user-images.githubusercontent.com/60436221/104686472-026f2680-5749-11eb-82e2-b0f3fba1ab9e.gif) |

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

aaronrobertshaw committed:
Add icon and bg color options to social links

aaronrobertshaw committed:
Update blocks using color support in theme-json doc

mkaz reviewed (approved):
This works well for me and I would agree I prefer this implementation over #25372

I didn't have any problems with implementation and usage, I would defer to @jasmussen for any comments on the CSS change, it looks like it just wraps it in a not if specified which worked fine in my testing, so I don't see an issue.

Good work on this, thanks! :+1:

jasmussen commented:
This is awesome. Theme developers are going to be so happy with this.

Here's what I see:

![Jan-13-2021 09-42-39](https://user-images.githubusercontent.com/1204802/104427804-fab25500-5583-11eb-91d0-a3512c40fec2.gif)

A couple of issues here:

**1.** I'm using TwentyTwentyOne Blocks and for whatever reason, setting the global color of the entire Social Icons block doesn't work. These appear to be the conflicting styles:
<img width="379" alt="Screenshot 2021-01-13 at 09 45 53" src="https://user-images.githubusercontent.com/1204802/104428000-33eac500-5584-11eb-8dc8-88ea9646a8fb.png">

You note: 

> some themes have specific styles for named background colors where it will override the selected icon color

We have to solve this. It'd be okay to increase specificity of the user-set colors. Setting the selector to this, solves it for me:

<img width="268" alt="Screenshot 2021-01-13 at 09 47 36" src="https://user-images.githubusercontent.com/1204802/104428146-672d5400-5584-11eb-9b62-201c6af3d49c.png">

Just add a CSS comment explaining why the extra specificity is added.

**2.** It's possible to override the color of just a single icon.

My very first instinct here is that this is a step too far. The color interface is identical whether you've selected the Social Icons parent block, or any individual block inside, meaning users will likely either set mixture of both, and be confused when one overrides the other. An additional instinct says that the use case of coloring only a single icon, as opposed to all of them at once, is an exceptionally rare use case that doesn't warrant the UX overhead.

I'm happy to defer to a sanity check by @jffng, @MaggieCabrera, @kjellr or @scruffian, who I believe have been yearning for this feature for a along time. If they prefer both the block itself, and individual social icons, to be colorable, I'll happily defer.

Thank you again for working on this 🏅

kjellr commented:
Thanks for the ping, @jasmussen! And thanks for the PR, @aaronrobertshaw! This is definitely a feature I'm excited to have. 🙌

I fully agree with @jasmussen's assessment, both about adding specificity to the user's color choice, and also the stance that these color options should apply only to _all_ social links.

aaronrobertshaw commented:
> 1. I'm using TwentyTwentyOne Blocks and for whatever reason, setting the global color of the entire Social Icons block doesn't work.

@jasmussen  I can't replicate the issue of the global color selection not being passed to the child Social Link blocks and being applied. The CSS styles in the screenshot also appear different than those I see.

| Demo | CSS |
|-------|-----|
| ![SocialLinks](https://user-images.githubusercontent.com/60436221/104671536-be205e00-5729-11eb-8a99-54c58d903601.gif) | <img width="423" alt="Screen Shot 2021-01-15 at 12 02 05 pm" src="https://user-images.githubusercontent.com/60436221/104671635-e0b27700-5729-11eb-86e2-d81b1c2d2aa9.png"> |

> > some themes have specific styles for named background colors where it will override the selected icon color
> 
> We have to solve this. It'd be okay to increase specificity of the user-set colors. Setting the selector to this, solves it for me:

I'll try and address this. The catch is the `has-<color>-color` and `has-<color>-background-color` styles come from the colors block support so I don't have control over their specificity. Especially if a theme is overwriting those exact same CSS classes as in the case I noted.

In the [previous PR](https://github.com/WordPress/gutenberg/pull/25372) I overcame this by actually determining the color value and applying it as an inline style. This lead to a somewhat more hacky approach and a lot of additional changing of code that others weren't a huge fan of.

> My very first instinct here is that this is a step too far. The color interface is identical whether you've selected the Social Icons parent block, or any individual block inside, meaning users will likely either set mixture of both, and be confused when one overrides the other. An additional instinct says that the use case of coloring only a single icon, as opposed to all of them at once, is an exceptionally rare use case that doesn't warrant the UX overhead.

That was my preference as well to only set the colors at the Social Links level as a group. Again, this was something that was covered in the [original PR](https://github.com/WordPress/gutenberg/pull/25372). 

The approach in this PR uses the colors block support to be able to apply the color CSS classes etc to the individual blocks. The controls added at the "group" or Social Links level are simply used to pass through values to set the individual blocks' color support attributes. I'm not sure if I can disable/hide the color controls at the individual block level but still retain the color block support behaviour. I'll look into it.

There's also a possibility that I can leverage some CSS variables to find some middle ground.

aaronrobertshaw committed:
Change to CSS variable approach

aaronrobertshaw commented:
I've updated this to use a new approach leveraging CSS variables set on the main Social Links block.

Improvements over the last iteration are:
- Theme styles for named colors no longer override user color selections
- No need to opt into the colors block support on inner social link blocks
- Color selections are made once only on the main social links block, simplifying UI

One downside was needing to store the color hex value for named colors in an attribute so the CSS variable inline style in save.js can still be correctly set. 

PR description has been updated to reflect current approach.

@jasmussen and @kjellr would you mind taking another look at this and seeing if it is more to your liking? 🙂

jasmussen commented:
Awesome work, still. The removal of the ability to color individual icons greatly simplifies the experience.

However I still can't get the colors to take:

<img width="1238" alt="Screenshot 2021-01-15 at 09 56 41" src="https://user-images.githubusercontent.com/1204802/104703784-0b93cf80-5718-11eb-92a4-f7f0d295cf07.png">

They need relatively high specificity. Which theme are you testing on?

Also, I'm not sure moving to CSS variables for these is a good idea. You might want two social links menus on a single page, with different color sets. For example you may have a light header with a couple icons at the end of your navigation menu, which are dark gray against the white background. Then you might also have social links showing up at the end of every post, with different colors. 

Won't the move to CSS variables make the choice of colors for one social icons block affect the colors of every other social icons blocks?

ntsekouras commented:
Hey 👋 - I'm wondering why you didn't try to implement with `color` supports in `block-json` in `SocialLink` block 🤔 

I think it would be much simpler and would need just the addition of a bit of css. I haven't checked it yet though, but I will as I am curious :)

aaronrobertshaw commented:
> They need relatively high specificity. Which theme are you testing on?

I tested with TwentyTwentyOne ([as per the last gif](https://github.com/WordPress/gutenberg/pull/28084#issuecomment-760604472)) using latest master, v1.0, and v1.1, TwentyTwenty, Maywood, and Varia. 

Prior to trying to replicate the issue you've seen, I updated all the themes as well. I also tested across browsers and still can't replicate the problem.

> Also, I'm not sure moving to CSS variables for these is a good idea. You might want two social links menus on a single page, with different color sets. 

> Won't the move to CSS variables make the choice of colors for one social icons block affect the colors of every other social icons blocks?

This is actually part of the beauty of the CSS variables approach, that will not be the case. [CSS variables can be inherited](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties#Inheritance_of_custom_properties) or scoped to a specific selector not just set globally on `:root`.

The current implementation I have in this PR sets the CSS variable as an inline style on the block. When the inner elements' CSS attempts to determine the variable's value via `var(--wp--social-links--icon-color)` for example, they look for the most locally scoped declaration of the CSS variable. In each block's case that would be the inline style for that individual block setting the variable.

Below are a few GIFs with multiple social links blocks using different color selections.

| 2021 | 2020 | Maywood | Seedlet | Varia |
|------|------|-----------|---------|------|
| ![SocialLinksColors2](https://user-images.githubusercontent.com/60436221/104861298-04203080-597b-11eb-96a4-7cfc28c86a14.gif) | ![SocialLinksColors](https://user-images.githubusercontent.com/60436221/104861304-08e4e480-597b-11eb-9e9f-a4f75baa8757.gif) | ![SocialLinksColors-Maywood](https://user-images.githubusercontent.com/60436221/104861308-0e422f00-597b-11eb-924a-d94c7106802b.gif) | ![SocialLinksColors-Seedlet](https://user-images.githubusercontent.com/60436221/104861313-1306e300-597b-11eb-987d-3eb2ae418a43.gif) | ![SocialLinksColors-Varia](https://user-images.githubusercontent.com/60436221/104861319-18642d80-597b-11eb-8065-d0112763100b.gif) |

aaronrobertshaw committed:
Prevent color selection loss switching theme

If switching themes, their color palettes may not have matching named colors.

In this situation, the previous color selection's named color would not match an option in the color controls so they would effectively clear the selection.

We are already maintaining a custom color value attribute to facilitate setting CSS variables in the save function. Using this instead of the withColors created color means if it does match we keep the selection. If it doesn't it is treated as a custom color.

aaronrobertshaw committed:
Use custom color attributes as fallback only

This partially undoes the prior commit that switched all color selections to use the custom attribute value. Instead, by using them as a fallback only, if the named color is also in the new theme's palette, that themes colors are used.

jasmussen commented:
I suspect the specificity issues are related to none of those themes being block based, and potentially not leveraging some of the global styles stuff that's underway. I can confirm it works in non block based themes, but falls apart in block based ones. You can try using the block based themes in https://github.com/WordPress/theme-experiments.

Here's my own:

<img width="1270" alt="Screenshot 2021-01-18 at 08 55 18" src="https://user-images.githubusercontent.com/1204802/104887487-a33d2c00-596b-11eb-94f5-788714dbdd48.png">

Here's Twenty Twenty One Blocks:

<img width="972" alt="Screenshot 2021-01-18 at 08 59 17" src="https://user-images.githubusercontent.com/1204802/104887502-a9cba380-596b-11eb-9d68-b5ed9dfe25cf.png">

Here's a video of a little inspecting:

https://cloudup.com/cxeNNecoxC0

As you mentioned, it works in vanilla Twenty Twenty One:

<img width="1456" alt="Screenshot 2021-01-18 at 09 07 07" src="https://user-images.githubusercontent.com/1204802/104888150-a1279d00-596c-11eb-9497-02cc9e791d39.png">

I can even see the right styles taking:

<img width="310" alt="Screenshot 2021-01-18 at 09 06 56" src="https://user-images.githubusercontent.com/1204802/104888165-a71d7e00-596c-11eb-93d0-7896c0d4fbc6.png">

☝️ that CSS appears to not be output at all in block based themes.

aaronrobertshaw commented:
@jasmussen Thanks for your help and patience on this one.

> I can confirm it works in non block based themes, but falls apart in block based ones. You can try using the block based themes in https://github.com/WordPress/theme-experiments.

I've tested with the themes from https://github.com/WordPress/theme-experiments including:
- Twenty Twenty One Blocks
- Twenty Twenty Blocks
- Ambitious
- Armando 
- Twenty Nineteen Blocks

All of them work fine for me with this PR (Twenty Nineteen Blocks has broken layout that's a separate issue).

The closest this PR now comes to Global Styles is using the `withColors` hook. That part is working for you or we wouldn't see the color controls. In your video I can see the inline styles setting the CSS variables as well. As you stated, it doesn't appear that you have the latest CSS containing the rules that use those.

The color selections via the CSS variables are applied to the individual Social Links [via simple CSS rules](https://github.com/WordPress/gutenberg/pull/28084/files?file-filters%5B%5D=.scss#diff-f799816d8bc9afdb9ba9be63ccbabf36aef4fc3f402fdf70236470f3a8c60c64R74) added to `packages/block-library/src/social-links/style.scss`

The CSS classes applied to the group social links block and the individual link blocks match between your video and what I see. Double checking the CSS rule, it should be applied so long as its in your stylesheet.

I've also cloned a fresh copy of Gutenberg, setup a new wp-env install, manually uploaded the block themes, rebuilt Gutenberg and re-tested. Same result, they all work for me. 

I'm obviously missing something.

| 2021 Blocks | 2020 Blocks | Ambitious | Armando |
|-------------|--------------|-----------|----------|
| ![SLC-TT1-blocks2](https://user-images.githubusercontent.com/60436221/104995077-6b201100-5a71-11eb-80fa-9f38811b3f2c.gif) | ![SLC-2020-blocks](https://user-images.githubusercontent.com/60436221/104995093-72dfb580-5a71-11eb-8692-7ef8d58f5069.gif) | ![SLC-Ambitious](https://user-images.githubusercontent.com/60436221/104995113-7a06c380-5a71-11eb-8cf5-e88fbc17d8a2.gif) | ![SLC-Armando](https://user-images.githubusercontent.com/60436221/104995123-80953b00-5a71-11eb-91bf-37ced4f677c4.gif) |

jasmussen commented:
My goodness.

I had `npm run dev` running in the background. Then from the main branch I checked out this one, and set the theme to TwentyTwentyOne, added a social links block and colored it, and it worked.

Then I opened a new tab, switched the theme to TwentyTwentyOne Blocks, opened the same post in the post editor, and saw that the colors _did not work_.

Then I debugged the two open tabs, without reloading, to figure out that the CSS to colorize each list item was simply missing in the tab that had TwentyTwentyOne Blocks. I even tried clearing cache thoroughly.

In a bit of frustration, I Ctrl+C'ed my `npm run dev` process and restarted it. And now it works fine in every theme.

I still don't quite understand why the existing background `npm run dev` compiled the files so that it worked in normal themes but not in block themes, or why restarting it fixed it also for block themes. But nevertheless, it appears to work fine. So I do apologize for the error, and thanks again for the work! Ship this one as fast as you can.

aaronrobertshaw commented:
> In a bit of frustration, I Ctrl+C'ed my `npm run dev` process and restarted it. And now it works fine in every theme.

On some rare occasions I've seen the my local watch task see the .scss files change but not build correctly when switching branches. I've not thought much of it at the time and have probably developed a bit of a habit of restarting `npm run dev` if I switch branches.

> I still don't quite understand why the existing background `npm run dev` compiled the files so that it worked in normal themes but not in block themes, or why restarting it fixed it also for block themes. But nevertheless, it appears to work fine. 

That one is a bit of a mystery!

> So I do apologize for the error, and thanks again for the work! Ship this one as fast as you can.

Not a problem. I'm glad it's working for you now. If nothing else, we've been pretty thorough testing this one across themes 🙂

jasmussen commented:
This one will make a lot of people happy.

As a followup, we should still make the change where we move the color options to the toolbar, as shown in [this comment](https://github.com/WordPress/gutenberg/issues/21605#issuecomment-692928358). Aaron if you have energy of course, otherwise CC: @mkaz? Alternatively we can open a new ticket to do just that, so it doesn't fall off the radar.

aaronrobertshaw commented:
> As a followup, we should still make the change where we move the color options to the toolbar, as shown in this comment. Aaron if you have energy of course, otherwise CC: @mkaz? Alternatively we can open a new ticket to do just that, so it doesn't fall off the radar.

How do you see the two different color options working in the block's toolbar? 

We have both icon color and icon background color. Do we need to combine those into a single toolbar menu?

jasmussen commented:
> How do you see the two different color options working in the block's toolbar?

Something along these lines:

<img width="1063" alt="Screenshot 2021-01-19 at 09 50 48" src="https://user-images.githubusercontent.com/1204802/105010319-d949e100-5a3b-11eb-8900-b81b332d53d5.png">

Some of this is a little dependant on work in https://github.com/WordPress/gutenberg/issues/27331 to further.

aaronrobertshaw commented:
> Some of this is a little dependant on work in #27331 to further.

Perhaps opening up a new ticket to move the controls into the block toolbar after https://github.com/WordPress/gutenberg/issues/27331 lands is the most efficient course of action? That is, unless others would rather see it done sooner. I don't have a strong opinion either way.

jasmussen commented:
I'll open a ticket.

nosolosw commented on docs/designers-developers/developers/themes/theme-json.md:
```diff
@@ -348,6 +348,7 @@ These are the current color properties supported by blocks:
 | Post Title | Yes | Yes | - | Yes |
 | Site Tagline | Yes | Yes | - | Yes |
 | Site Title | Yes | Yes | - | Yes |
+| Social Links | Yes | - | - | Yes |
```
:wave: While this PR added support for icon color & background, it doesn't use the block support mechanism, hence these properties can't be targeted via theme.json. I've prepared a fix for this at https://github.com/WordPress/gutenberg/pull/29294

aaronrobertshaw replied:
Thank you for the fix!

nosolosw commented:
:wave: I've filed an issue at https://github.com/WordPress/gutenberg/issues/29297 that relates to the use of CSS Custom Properties in this PR.
